pagefaultgames/pokerogue
# [QoL] Don't offer ball rewards when you have max in Classic mode

## What are the changes?

Balls are no longer offered in rewards if you have max of them in classic.

## Why am I doing these changes?



Because it's stupid if you have 99 Great Balls and the rewards gives you Great Balls and you can take it and not get anything because you already have max so you just waste your reward for that floor.

## What did change?


I added a check in the modifierPool that sets the weights of a ball to 0 if you have 99 (or more) of that ball and you're in Classic.

### Screenshots/Videos


How it currently works:
![image](https://github.com/pagefaultgames/pokerogue/assets/36677462/f7468d37-5382-4724-8d75-e66930f07f14)
Wave 1, Red Ball in rewards
![image](https://github.com/pagefaultgames/pokerogue/assets/36677462/34cf681c-8da3-4398-a0bb-17ed48051fc4)
Wave 2, Too many balls
![image](https://github.com/pagefaultgames/pokerogue/assets/36677462/64c7048a-ab1d-43ae-92ea-f54751edb1c2)
Wave 2, no Red Balls in rewards
![image](https://github.com/pagefaultgames/pokerogue/assets/36677462/2703f8bd-08a4-4da0-947f-1f61c16f717e)
Wave 2, no Red Balls after 14 rerolls with 99 in bag
![image](https://github.com/pagefaultgames/pokerogue/assets/36677462/bbfb5e9f-4ff3-47cd-9c11-b3d092db3f6a)
I got into a double battle with 2 flying types after that so I had to restart but here it is still offering Red Ball when I have 98
![image](https://github.com/pagefaultgames/pokerogue/assets/36677462/487adf58-7022-4acb-b88d-511d51e9bc42)
Change doesn't apply in Endless:
![image](https://github.com/pagefaultgames/pokerogue/assets/36677462/005cbf86-6411-4f14-bdce-2ac9e037319b)

## How to test the changes?




In src/modifier/modifier-type.ts, change the following line:
`export const modifierTypes = {
  POKEBALL: () => new AddPokeballModifierType("pb", PokeballType.POKEBALL, 5),`
The last parameter in AddPokeballModifierType, currently 5, is the amount of balls you get from one reward. Change it to 94 or more so you can hit 99 after the first wave.
In the same file, in modifierPool:
`const modifierPool: ModifierPool = {
  [ModifierTier.COMMON]: [
    new WeightedModifierType(modifierTypes.POKEBALL, (party: Pokemon[]) => (party[0].scene.gameMode.isClassic && party[0].scene.pokeballCounts[PokeballType.POKEBALL] >= 99) ? 0 : 6, 6),`
Change the two 6's at the end (which are the weight and the maximum weight, respectively) to something really high so you can get 99 ASAP and ensure that luck isn't a factor when you're testing if it was removed from the pool.

Then in phases.ts, at the end of SelectModifierPhase.getRerollCost:
`return Math.min(Math.ceil(this.scene.currentBattle.waveIndex / 10) * baseValue * Math.pow(2, this.rerollCount), Number.MAX_SAFE_INTEGER);`

Change Number.MAX_SAFE_INTEGER to 1, or just make the whole line `return 1`. This will set the reroll cost to 1 so you can test rerolling many times.

Then start a new run, get to 99 red balls and bear witness to its absence in the next set of rewards.

## Checklist
